### ISSUE TITLE
With "nlb-ip", we never see targets go to "unhealthy" even with unhealthy targets. Targets stay in "draining" status.

### ISSUE THREAD

**Issue by keperry** [kubernetes-sigs/aws-load-balancer-controller#1633]
aws-load-balancer-controller: v2.0.0
eks version: 1.17
service: node-group with "nlb-ip" for loadbalancer type

With "nlb-ip", we never see targets go to "unhealthy" even with unhealthy targets.  Our pods take our main traffic via port 8443 (tcp traffic), but handle a healthcheck on port 8082 with a path of /healthcheck.  

Our annotations look like this:
```
    service.beta.kubernetes.io/aws-load-balancer-type: "nlb-ip"
    service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: "*"
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-healthy-threshold: "2"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-interval: "10"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-path: /healthcheck
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-port: "8082"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-protocol: HTTP
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-timeout: "6"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-unhealthy-threshold: "2"
    service.beta.kubernetes.io/aws-load-balancer-connection-draining-timeout: "60"
    service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
    service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "60"
```
Our pods handle graceful shutdown, once they go into "Terminating" status, they wait for a configurable period of time and signal to our readiness route (/healthcheck on port 8082) to return a 500.  While the pods show "not-ready" in kube, they always only show as "draining" in the target group, and even though the pod is configured to shut down after 2.5 minutes, they never show as unhealthy.  This results in traffic still being sent to the "draining' target.  

Upon testing alb targets, they seem to update to "unhealthy" almost immediately when the pod itself is "not-ready" in kube.  Is there something else we can tweak or is there a bug in how this is operating with "nlb-ip" type targets.

Overall, appreciate all the work done here, it really looks promising for nlb's going forward.

**Comment by kishorj** [comment]
Are you using similar deployment specs for ALB and NLB targets? The targets do transition to unhealthy state - try using a different port than 8082, or the healthcheck service return non 200-399 HTTP codes. It looks like the targets got de-registered before transitioning to unhealthy.

This annotation does not have any effect
`    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: tcp`

These annotations gets ignored for NLB 
```    
     service.beta.kubernetes.io/aws-load-balancer-connection-draining-timeout: "60"
    service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "60"
```
Connection draining timeout will be configurable once the support for target group attributes get merged PR #1632. The current default value for deregistration delay is 300s. 
NLB does not support idle timeout yet.

**Comment by keperry** [comment]
@kishorj - yes, we are using similar deployment specs for both targets.  afaict, the targets never transition to an "unhealthy" state.  They go immediately to "draining" and stay that way for the entire "deregistration delay".  Our healthcheck route does already return a 500 http error code during shutdown.  Not sure how a different port would help.  The targets don't seem to have an issue hitting the healthcheck on the pod (they remain healthy when they are up and running).

We have tried removing our readiness probe, but the as soon as the pod goes to terminating, the controller appears to deregister.  Again, this appears to behave differently on the alb side of things.  We see our pod go to unhealthy almost immediately. 

Another additional thing that would be helpful is to have a similar thing like `aws-api-debug` flag so we can see all the calls.
Can you think of another way to test this?  I am digging through the controller code as well.

**Comment by kishorj** [comment]
@keperry I want to clarify on using a different port - I didn't intend to say that it would fix the issue, I just wanted to point out if a different port were used for healthcheck, the targets should be unhealthy.

NLB and ALB are different implementations and have different behaviors.

I will dig further and get back with more details.

**Comment by keperry** [comment]
@kishorj - thanks for your help!  I think I have verified that this is "expected behavior."  
If I setup NLB with targets manually.  When I manually kill the pod, and then the readiness probe fails, I do see the healthcheck starts failing in the target group, it will go to unhealthy state in the target group.  If I kill the pod, and then deregister the pod manually, I never see the pod go to "unhealthy."   

Note:  when I retested this on alb, it appears to follow the same behavior as nlb - target goes to draining very quickly, and then never went to "unhealthy."
